feat: Adding ability to configure yamux rpc connections multiplexor (#25466)
hashicorp/nomad: website/content/docs/configuration/rpc.mdx

---
layout: docs
page_title: rpc Block in Agent Configuration
description: >-

>    Configure Nomad's RPC behavior in the `rpc` block of a Nomad agent configuration. Modify RPC session configuration, change TPC backlog, keepalive interval, and open socket timeouts.

Reviewer: Oops, typo: 

```suggestion
   Configure Nomad's RPC behavior in the `rpc` block of a Nomad agent configuration. Modify RPC session configuration, change stream accept backlog, keepalive interval, and open stream timeouts.
```

---

# rpc Block in Agent Configuration

<Placement groups={['rpc']} />

> This page provides reference information for configuring Nomad's RPC behavior in the `rpc` block of a Nomad agent configuration. Modify RPC session configuration, change TPC backlog, keepalive interval, and open socket timeouts.

Reviewer: As noted in my earlier comments, I want to make it clear these aren't intended for typical use. Pretty much the first thing I'd do if I saw a bug reported from someone configuring with these knobs is to assume they're messing around with something they don't understand and tell them to switch it back. :grin: So I'm going to go a little overboard here in fact:

```suggestion
This page provides reference information for configuring Nomad's RPC behavior in
the `rpc` block of a Nomad agent configuration. Modify RPC session
configuration, stream accept backlog, keepalive interval, and open stream
timeouts. We strongly recommend leaving these values unconfigured to use the
default values, which have been production-proven on clusters of thousands of
nodes. You should only configure the `rpc` block if you have specific reason to
believe it will improve your specific use case.
```

Another item here is that I'm realizing you're saying "TCP backlog", but that's not what the `accept_backlog` configures at all -- it configures the _stream_ accept backlog for the session.

```hcl
rpc {
  accept_backlog           = 256
  keep_alive_interval      = "30s"
  connection_write_timeout = "10s"
  max_stream_window_size   = 262144
  stream_open_timeout      = "75s"
  stream_close_timeout     = "5m"
}
```

## `rpc` Parameters

> - `accept_backlog` `(int: 256)` - AcceptBacklog is used to limit how many streams may be waiting an accept.

Reviewer: We're copying these descriptions right out of the library docstrings, but the audience for those docstrings are consumers of the yamux library, not Nomad cluster admininstrators. For example:

```suggestion
- `accept_backlog` `(int: 256)` - The accept backlog limits how many RPC streams (requests) can be waiting to be accepted by the RPC server.
```

Take a pass over these rest of these fields and apply the same approach.

  (requests) can be waiting for acceptance by the RPC server.

- `keep_alive_interval` `(string: 30s)` - How often to perform the keep alive

- `connection_write_timeout` `(string: 10s)` - Safety valve timeout after which
  you should suspect a problem with the underlying connection and close it. This
  is only applied to writes, where there is generally an expectation that things
  move along quickly.

> - `max_stream_window_size` `(int: 262144)` - Controls the maximum window size
>   allowed for a stream.

Reviewer: ```suggestion
- `max_stream_window_size` `(int: 262144)` - Controls the maximum window size
  allowed for a stream, in bytes.
```

> - `accept_backlog` `(int: 256)` - The accept backlog limits how many RPC streams (requests) can be waiting to be accepted by the RPC server.
> 
> - `keep_alive_interval` `(string: 30s)` - KeepAliveInterval is how often to perform the keep alive
> 
> - `connection_write_timeout` `(string: 10s)` - ConnectionWriteTimeout is meant to be a "safety valve" timeout after
>   we which will suspect a problem with the underlying connection and
>   close it. This is only applied to writes, where's there's generally
>   an expectation that things will move along quickly.
> 
> - `max_stream_window_size` `(int: 262144)` - MaxStreamWindowSize is used to control the maximum window size
>   that we allow for a stream.
> 
> - `stream_open_timeout` `(string: 75s)` - StreamOpenTimeout is the maximum amount of time that a stream will be
>   allowed to remain in pending state while waiting for an ack from the peer.
>   Once the timeout is reached the session will be gracefully closed.
>   A zero value disables the StreamOpenTimeout allowing unbounded blocking on OpenStream calls.
> 
> - `stream_close_timeout` `(string: 5m)` - StreamCloseTimeout is the maximum time that a stream will allowed to
>   be in a half-closed state when `Close` is called before forcibly
>   closing the connection. Forcibly closed connections will empty the
>   receive buffer, drop any future packets received for that stream,
>   and send a RST to the remote side.

Reviewer 2: ```suggestion
- `accept_backlog` `(int: 256)` - Limits how many RPC streams
  (requests) can be waiting for acceptance by the RPC server.

- `keep_alive_interval` `(string: 30s)` - How often to perform the keep alive

- `connection_write_timeout` `(string: 10s)` - Safety valve timeout after which
  you should suspect a problem with the underlying connection and close it. This
  is only applied to writes, where there is generally an expectation that things
  move along quickly.

- `max_stream_window_size` `(int: 262144)` - Controls the maximum window size
  allowed for a stream.

- `stream_open_timeout` `(string: 75s)` - Maximum amount of time that a stream
  remains in pending state while waiting for an ACK from the peer. Once the
  timeout is reached, Nomad gracefully closes the session. A zero value disables
  the stream open timeout, allowing unbounded blocking on open stream calls.

- `stream_close_timeout` `(string: 5m)` - When `close` is called, maximum time
  that a stream is in a half-closed state before forcibly closing the
  connection. Forcibly closed connections empty the receive buffer, drop any
  future packets received for that stream, and send an RST to the remote side.
```

Some suggestions so the text matches our docs style guide.
